src/BTree.mo

```diff
@@ -0,0 +1,110 @@
+/// Imperative sequences as B-Trees.
+
+import A "Array";
+import I "Iter";
+import List "List";
+import Option "Option";
+import Order "Order";
+import P "Prelude";
+import Prim "mo:⛔";
+
+module {
+
+  /// Constants we use to shape the tree.
+  /// See https://en.wikipedia.org/wiki/B-tree#Definition
+  module Constants {
+    let MAX_CHILDREN = 4;
+  };
+
+  public type Compare<K> = {
+    compare : (K, K) -> Order.Order
+  };
+
+  public type Data<K, V> = [(K, V)];
+
+  public type Index<K, V> = {
+    data : Data<K, V>;
+    trees : [Tree<K, V>];
+  };
+
+  public type Tree<K, V> = {
+    #index : Index<K, V>;
+    #data : Data<K, V>;
+  };
+
+  func find_data<K, V>(data : Data<K, V>, find_k : K, c : Compare<K>) : ?V {
+    for ((k, v) in data.vals()) {
+      if (c.compare(k, find_k) == #equal) { return ?v };
+    };
+    return null
+  };
+
+  func find<K, V>(t : Tree<K, V>, k : K, c : Compare<K>) : ?V {
+    switch t {
+      case (#data(d)) { return find_data<K, V>(d, k, c) };
+      case (#index(i)) {
+        for (j in I.range(0, i.data.size())) {
+          switch (c.compare(k, i.data[j].0)) {
+            case (#equal) { return ?i.data[j].1 };
+            case (#less) { return find<K, V>(i.trees[j], k, c) };
+            case _ { }
```

Comment:
Since `trees` is an ordered array, this could be a great place to do a modified binary search on the `i.trees` array to find the correct subtree to navigate down.

Reply:
Also, are there any (scalability) concerns about the call stack size since we're using recursion vs. a loop?

Reply:
> this could be a great place to do a modified binary search on the i.trees array to find the correct subtree to navigate down.

Yes, true!

OTOH, that would not be a "baby step".  : )

Ideally, we'd profile both variations, at various choices of `m` (the branching factor for B-Trees seems to be called `m` by everyone for some reason) and we'd have some way to know if there's a benefit, and how much.  There may be little benefit when `m` is small, or it may even be worse (?)

Reply:
> Also, are there any (scalability) concerns about the call stack size since we're using recursion vs. a loop?

[Within the compiler, tail recursion to the same function gets transformed into a loop, with no stack/recursion.](https://github.com/dfinity/motoko/blob/c2bf42fed86259593d5930c0edea4947ada13663/src/ir_passes/tailcall.ml#L17)

Reply:
It might end up being the case that a BTree with index/sub-tree size 16/32/64 removes this issue since the tree is more shallow and there's less rebalancing steps required 🤷‍♂️ 

Just for context, the reason the insertion performance is important to me is that I'd prefer to find a balanced data structure solution that avoids hitting the message instruction limit issues without relying on deterministic time slicing.

Reply:
Yep, I think B-Trees will be shallow for any realistic internal node size (16, 32, etc.).

As mentioned above, in Motoko, any self-contained tail-recursive function (not two mutually-recursive ones) will become a single while loop without a stack during compilation.

For any realistic BTree, that stack-free loop should run at most a handful of iterations (less than 10), so the cycle limit will certainly not matter for searching, even if you need to search for a bunch of records in random positions all within one message response time.